man/mib_api.3
.TH MIB_API 3 "18 Dec 1999"
.UC 5
.SH NAME
init_mib, add_mibdir, init_mib_internals,
read_module, read_mib, read_all_mibs,
read_objid, read_module_node,
get_module_node,
snmp_set_mib_warnings, snmp_set_save_descriptions,
print_mib,
print_variable, print_value, print_objid, print_description - mib_api functions
.SH SYNOPSIS
.B "#include <stdio.h>
.br
.B "#include <sys/types.h>
.br
.B "#include <ucd-snmp/asn1.h>
.br
.B #include <ucd-snmp/mib.h>
.br
.B #include <ucd-snmp/parse.h>
.PP
.B "void init_mib( void )
.br
.B "int add_mibdir( char *dirname )
.br
.B "int add_module_replacement( char *old_module, char *new_module, char *tag, int len )
.br
.B "void init_mib_internals( void )
.br
.B "struct tree *read_module( char *name )
.br
.B "struct tree *read_mib( char *filename )
.br
.B "struct tree *read_all_mibs( void )
.PP
.B "void print_mib( FILE *fp)
.PP
.B "int read_objid( char *input, oid *output, int *out_len)
.br
.B "int get_module_node( char *name, char *module, oid *objid, int *objidlen )
.PP
.B "void print_variable( oid *objid, int objidlen, struct variable_list *variable)
.br
.B "void print_value( oid *objid, int objidlen, struct variable_list *variable)
.br
.B "void print_objid( oid *objid, int objidlen);
.br
.B "void print_description( oid *objid, int objidlen);
.br
.B "void snmp_set_mib_warnings( int );
.br
.B "void snmp_set_save_descriptions( int );
.PP
.SH DESCRIPTION
The functions dealing with MIB modules fall into four groups.
Those dealing with initialisation, those that read in and parse MIB files,
those that search the MIB tree, and various output routines.

.I init_mib
is a convenience function that handles all calls to
.I add_mibdir, read_module
and
.I read_mib
for standard applications.
It should be called before any other routine that manipulates or accesses the
MIB tree.  This routine sets up various internal structures, as well as
reading in the default MIB modules, as detailed below.

.I add_mibdir
is used to define the range of directory locations which are searched for files
containing MIB modules (one module per file).
By default, this will be set to the directory
.I DATADIR/mibs
but this can be overridden by setting the environment variable
.I MIBDIRS
to a (colon-separated) list of directories to search.
Note that this does not actually load the MIB modules located
in that directory, but is an initialisation step to make them available.
This function returns a count of files found in the directory, or a -1
if there is an error.  

.I init_mib_internals
sets up the internal structures, preparatory to reading in MIB modules.
It should be called after all calls to
.IR add_mibdir ,
and before and calls to
.IR read_module .
This is called automatically if
.I init_mib
is used.

.I add_module_replacement
can be used to allow new MIB modules to obsolete older ones, without
needing to amend the imports clauses of other modules.
It takes the names of the old and new modules, together with an indication
of which portions of the old module are affected.
.RS
.TS
tab(+);
lb lb lb
l  l  l.
tag + len + load the new module when:
NULL + 0 + always (the old module is a strict subset of the new)
name + 0 + for the given tag only
name + non-0 + for any identifier with this prefix
.TE
.RE
It can also be used to handle errors in the module identifiers used
in MIB import clauses (such as referring to
.I RFC1213
instead of
.I RFC1213-MIB
).

.I read_module
locates and
parses the module specified, together with any modules that it imports
from, and adds the contents of these modules to the active MIB tree.
Note that
.I add_mibdir
must first be called to add the directory containing the file with the
module definition, if this is not in the standard path.
.br
By default, the following mib modules will be loaded:  IP-MIB, IF-MIB,
TCP-MIB, UDP-MIB, SNMPv2-MIB, RFC1213-MIB, UCD-SNMP-MIB.
This can be overridden by setting the environment variable
.I MIBS
to a (colon-separated) list of modules to load.
If this variable starts with a plus character, then the specified modules
are added to the default list.  Otherwise only those modules listed are
loaded (together with any others they import from).
If
.I MIBS
is set to
.IR ALL ,
.I read_all_mibs
is called to load all the MIB files found in all the specified
.IR MIBDIRS .


.I read_mib
parses the file specified, together with any modules that it imports
from, and adds the contents to the active MIB tree.
Such a file can contain more then one module, though care must be taken
that any imports occur earlier in the file, if they are not to be read
from the installed modules.
Note that the file specified does not need to be in any of the
directories initialised by
.I add_mibdir
(or the default setup), though any imported modules do.
.br
The environment variable
.I MIBFILES
can be set to a (colon-separated) list of files containing mibs to load.

.I read_objid
takes a string containing a textual version of an object identifier
(in either numeric or descriptor form), and transforms this into
the corresponding list of sub-identifiers.  This is returned in the
.I output
parameter, with the number of sub-identifiers returned via
.IR out_len .
When called, out_len must hold the maximum length of the
.I output
array.
This function returns a value of 1 if it succeeds in parsing the string
and 0 otherwise.

.I get_module_node
takes a descriptor and the name of a module, and returns the corresponding
oid list, in the same way as
.I read_objid
above.
.br
If the module name is specified as "ANY", then this routine will assume
that the descriptor given is unique within the tree, and will return the
matching entry.  If this assumption is invalid, then the behaviour
as to which variable is returned is implementation dependent.

.I print_mib
will print out a representation of the currently active MIB tree to
the specified FILE pointer.

.I print_variable
will take an object identifier (as returned by
.I read_objid
or
.I get_module_node
) and an instance of such a variable, and prints out
the textual form of the object identifier together with the value
of the variable.
.br
A related routine
.I sprint_variable
takes an initial parameter of a string buffer, in which to print this
information.

.I print_value
and
.I sprint_value
do the same as the equivalent
.I print_variable
routines, but only displaying the value of the variable, without
the corresponding object identifier>

.I print_objid
and
.I sprint_objid
.br
take an object identifier (without an accompanying variable instance)
and print out the textual representation.
.br
Note that identifiers lying within the mib-2, experimental
or private subtrees are listed from this point onwards,
rather than with the full object identifier.

.I print_description
takes an object identifier (as for
.I print_objid
above) and prints out the associated description.
.b
Note that there is no corresponding routine
.I sprint_description

By default the parser does not save descriptions - they may be huge.
In order to be able to print them, you must call
.IR "snmp_set_save_descriptions(1)" .

In general the parser is silent about what strangenesses it sees
in the mib files. To get warnings reported, call
.I snmp_set_mib_warnings
with a parameter of \fI1\fR (or \fI2\fR for even more warnings).

.SH "ENVIRONMENT VARIABLES"
.TP 10
MIBDIRS
A colon separated list of directories to search for MIB modules.
Default: LIBDIR/snmp/mibs
.TP 10
MIBFILES
A colon separated list of files to load.
Default: (none)
.TP 10
MIBS
A colon separated list of MIB modules to load.
Default: IP-MIB:IF-MIB:TCP-MIB:UDP-MIB:SNMPv2-MIB:RFC1213-MIB:UCD-SNMP-MIB.
.SH "SEE ALSO"
snmp_api(3)


NEWS
This file contains major changes in released revisions.  Please read
the ChangeLog file for a more detailed listing of changes/bug fixes.

*5.1.3.2*
 Ports:
   Win32:
     - patch 1258146: MinGW/Cygwin winservicerc.lo compile error
     - Fix for 'select: No such file or directory' in snmpdelta

*5.1.3.1*
 Fixes:
   - Fix premature close of file descriptors when run in the background
   - Fix compile/link problems when configured with --enable-mini-agent
   - Fix compiling Win32 without the Platform SDK (PSDK)

*5.1.3*
 Fixes:
   security:
     - fix potential race condition in fixproc script
     - fix DOS vulnerability on tcp connections

   agent:
     - bug 1034008: memory leak using SET for table_dataset
     - patch 1052460: fix agent deadlock on exec
     - bug 1055781: get-next fails to step into interfaces group correctly
     - bug 1056760: agent ignores ifspeed, type settings in snmpd.conf
     - Persistent files in directory defined by snmp.conf persistentDir were 
       not being loaded at startup
     - 1062986: pass and pass_persist fail and crash snmpd
     - patch 1052460: agent deadlock on exec
     - fix bug 1056760: agent ignores ifspeed, type settings in snmpd.conf
     - bug 119106, ipAdEntIfIndex is wrong
     - bug 986238: snmpd loops forever
     - bug 615744: Spurious DISMAN-EVENT traps
     - patch 1040718: Agentx error propagation and infinite loop
     - fix error handling for proxy get-next requests

   snmptrapd:
     - new configure option to exclude AgentX subagent code
     - new runtime option to exclude table registrations 

   library:
     - process pre-mib config tokens in optional config files at the right time
     - get rid of strtok (patch 1040330, backported by Thomas Anders, fixes 
       bug 1040686)
     - consistent handling of '+' for MIB and MIB directory handling from all
       sources (config file, environment variables, command line)
     - handle agentXsocket token in sub-agent configuration files
     - several AgentX fixes

 Ports:
   Linux:
     - use ethtool ioctl to detect gigibit interface speeds
     - Fix reversed sysIORawSent/Received
     - 64bit fixes to interface and ssRawCpu statistics
     - integrate fixes from RedHat and Debian

   Tru64:
     - build fixes; README.tru64 added

   FreeBSD:
     - apply patch 1056927: 5.2-p03: freebsd interface bugs
     - fix bug 1055781: get-next fails to step into interfaces group correctly

   NetBSD:
     - integragte fixes from NetBSD port

   Win32:
     - Cygwin compiler fixes
     - bug 926389: Win32 event log logging


*5.1.2*
 New:
   - Minor improvements to snmpdelta (error reporting)
   - Minor improvements to snmpnetstat (IPv6 output)

 Fixes:
   - Misc 64bit processor fixes.
   - Misc perl build and install fixes.

 Ports:
   - Linux 2.6 improvements
   - Win32 Suport for TCP and UDP over IPv6 via Winsock version 2
   - Win32 fixes in many places.
   - Win32 service support for snmptrapd
   - Win32 support for snmpconf

*5.1.1*
 New:
   - test suite supports testing over other transports (tcp, udp6, unix, ...)
     (see the -P switch to the testing/RUNTESTS script)
   - Solaris supports the use of it's PKCS#11 library for supporting
     cryptographic functions (OpenSSL isn't required if PKCS#11 is available)
     (see configure's --with-pkcs flag)

 Fixes:
   - Improvements on 64 bit architectures.
   - A few minor memory leaks fixed.
   - An extremely large number of minor bug fixes.
   - Many perl module specific bug fixes.
   - snmpd will safely handle more signals.

 Ports:
   - Many many significant Windows improvements.
     - AgentX support is working again under windows.
     - SCM support is built into the agent to allow the agent to be
       started as a windows service.
     - MSDOS names are supported in path names.
     - A win32 build script in win32/build.pl
     - Support for the MinGW compiler
     - (see the README.win32 file for details on new ports)
     - Various helpful win32/*.bat files for installation, etc.
   - Some linux 2.6 support improvements

*5.1*
 New:

    Building:
      - configure is now generated using autoconf 2.57.
      - The make system now supports "make uninstall"
      - configure & make supports a --with-install-prefix option
      - new configure options to disable building of the agent, apps, ...
        ./configure --help for a list of package sections that can be excluded.
      - new configure option to remove specific code pieces (mib reading, ...)
        ./configure --help for a list of code areas that can be excluded.

    snmpd:
      - linkUp and linkDown notifications can be enabled
        (see the linkUpDownNotifications snmpd.conf token documentation)
      - notifications (traps) can be generated for disman/monitor events.
        (see the snmpd.conf documentation on "monitor" and "notificatonEvent")
      - new snmpd.conf tokens:  includeAllDisks
      - the UCD-DISKIO-MIB will now work on linux.
      - com2sec mappings for SNMP over unix sockets.
      - some speed improvements have been made that should increase the
        speed of the agent's processing time.
      - for mib code using the table iterator, an auto-caching mechanism
        which can greatly speed up access at the cost of caching.
        (see "stash_cache" in the injectHandler token docs in snmpd.conf(5))
      - Some of the agent modules were completely rewritten (and
        should be faster and cleaner).
      - A number of new APIs and helper modules are available, but
        there is no major summary to easily list.
      - A number of new Net-SNMP specific MIBs now exist.

    snmptrapd:
      - snmptrapd now supports forwarding of notifications.
        (see the snmptrapd.conf manual for details)
      - snmptrapd's new -t switch can be used to disable syslog.
        (useful for a no-output snmptrapd with only traphandle support)

    apps:
      - snmptable sports 3 new options: fixed width, max-repeaters,

snmplib/mib/mib_api.c
/*******************************
 *
 *	mib_api.c
 *
 *	Net-SNMP library - MIB-handling interface
 *
 *	General routines
 *
 *******************************/

#include <config.h>

#if HAVE_STDLIB_H
#include <stdlib.h>
#endif
#if HAVE_STRING_H
#include <string.h>
#else
#include <strings.h>
#endif
#include <sys/types.h>

#include <net-snmp/mib_api.h>


		/*
		 * Temporary stuff....
		 */
int tags_match( char *taglist, const char *tag );

extern char *confmibdir;
extern char *confmibs;

char *conf_read_string( char *tag )
{
    char  *env_var;

    if ( strcmp( tag, "mibdirs" ) == 0 ) {
	env_var = getenv("MIBDIRS");

	if ( env_var != NULL ) {
	     return strdup( env_var );
	}
	else if ( confmibdir != NULL ) {
	     return strdup( confmibdir );
	}
	else {
	     return strdup( DEFAULT_MIBDIRS );
	}
    }

    if ( strcmp( tag, "mibs" ) == 0 ) {
	env_var = getenv("MIBS");

	if ( env_var != NULL ) {
	     return strdup( env_var );
	}
	else if ( confmibdir != NULL ) {
	     return strdup( confmibs );
	}
	else {
	     return strdup( DEFAULT_MIBS );
	}
    }

    return NULL;
}


		/**************************************
		 *
		 *	Public API
		 *	   (see <net-snmp/mib_api.h>)
		 *
		 **************************************/
		/** @package mib_api */


   /**
    *
    *  Initialise the MIB-handling subsystem
    *
    *  Returns 0 if successful, -ve otherwise
    *
    */
int mib_init( char *tags )
{
    char *cp;

    if ( !tags_match( tags, "mib" )) {
	return -1;
    }

    (void)smiExit( );
    (void)smiInit( NULL );

	/*
	 * ToDo:  Handle this via config settings instead
	 */

		/*  Set up the list of directories to search */
    cp = conf_read_string( "mibdirs" );
    if ( cp != NULL ) {
	mib_set_directories( cp );
	free( cp );
    }

		/*  Read in the appropriate MIB modules */
    cp = conf_read_string( "mibs" );
    if ( cp != NULL ) {
	if ( strcmp( cp, "ALL" ) == 0 ) {
	    (void)mib_load_all();
	}
	else {    
	    (void)mib_load_modules( cp );
	}
	free( cp );
    }

    return 0;
}


   /**
    *
    *  Close down the MIB-handling subsystem
    *
    *  Returns 0 if successful, -ve otherwise
    *
    */
int mib_close_down( char *tags )
{
    if ( !tags_match( tags, "mib" )) {
	return -1;
    }

    (void)smiExit( );
    (void)mib_set_directories( NULL );	/* Clear the directory list */


    return 0;
}



		/**************************************
		 *
		 *	internal utility routines
		 *
		 **************************************/
		/** @package mib internals */

   /*
    *  int tags_match()
    *
    *  Should this initialisation routine be called?
    *
    *  Returns 1 if true, 0 if not
    */
#define YES 1
#define NO  0

#ifndef PATH_MAX
#define PATH_MAX 512
#endif

	/*
	 * XXX - this belongs in a more general utility subsystem
	 */
int tags_match( char *taglist, const char *tag )
{
    char listbuf[ PATH_MAX ];
    char tagbuf[  PATH_MAX ];

			/* Nothing to check - default to YES */
    if (( taglist == NULL ) || ( *taglist == '\0' )) {
	return YES;
    }

			/* List of subsystems to skip */
    if (( *taglist == '!' ) || ( *taglist == '-' )) {
	return 1-tags_match( taglist+1, tag );
    }

			/* Too many tags to check, so assume OK */
    if (strlen(taglist)+2 >= PATH_MAX ) {
	return YES;
    }

			/* This tag is too long to check, so fail */
    if (strlen(tag)+2 >= PATH_MAX ) {
	return NO;
    }

    snprintf( tagbuf,  PATH_MAX, ",%s,", tag );
    snprintf( listbuf, PATH_MAX, ",%s,", taglist );

    if ( strstr( listbuf, tagbuf ) != NULL ) {
	return YES;
    }
    else {
	return NO;
    }
}
